include/nanobind/eigen/sparse.h

```diff
@@ -0,0 +1,124 @@
+    using StorageIndexTensor = tensor<StorageIndex, shape<any>, c_contig, device::cpu>;
+
+    using ScalarCaster = make_caster<ScalarTensor>;
+    using StorageIndexCaster = make_caster<StorageIndexTensor>;
+
+    NB_TYPE_CASTER(T, const_name<row_major>("scipy.sparse.csr_matrix[",
+                                            "scipy.sparse.csc_matrix[")
+                   + make_caster<Scalar>::Name + const_name("]"));
+
+    ScalarCaster scalar_caster;
+    StorageIndexCaster storageindex_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+        auto obj = borrow<object>(src);
+        try {
+            object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+            if (!obj.type().is(matrix_type))
+                obj = matrix_type(obj);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        if (object data_o = obj.attr("data"); !scalar_caster.from_python(data_o, flags, cleanup))
+            return false;
+        ScalarTensor& values = scalar_caster.value;
+
+        if (object indices_o = obj.attr("indices"); !storageindex_caster.from_python(indices_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& inner_indices = storageindex_caster.value;
+
+        if (object indptr_o = obj.attr("indptr"); !scalar_caster.from_python(indptr_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& outer_indices = storageindex_caster.value;
+
+        object shape_o = obj.attr("shape"), nnz_o = obj.attr("nnz");
+        Index rows, cols, nnz;
+        try {
+            if (len(shape_o) != 2)
+                return false;
+            rows = cast<Index>(shape_o[0]);
+            cols = cast<Index>(shape_o[1]);
+            nnz = cast<Index>(nnz_o);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        value = EigenMapSparseMatrix<Scalar,
+                                     T::Flags &(Eigen::RowMajor | Eigen::ColMajor),
+                                     StorageIndex>(rows,
+                                                   cols,
+                                                   nnz,
+                                                   outer_indices.data(),
+                                                   inner_indices.data(),
+                                                   values.data());
+
+        return true;
+    }
+
+    static handle from_cpp(T &&v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        if (policy == rv_policy::automatic ||
+            policy == rv_policy::automatic_reference)
+            policy = rv_policy::move;
+
+        return from_cpp((const T &) v, policy, cleanup);
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        // FIXME: Looks like UB. Is there anything that can be done about this?
```

Comment:
This indeed doesn't seem like a good design. I suggest the following instead:

Option 1: give up (untested)

```cpp
if (!v.isCompressed()) {
    PyErr_SetString(PyExc_ValueError, "nanobind: unable to return an Eigen sparse matrix that is not in a compressed format. Please call `.makeCompressed()` before returning the value on the C++ end.");
    return handle();
}
```

Option 2: copy to turn the matrix into a compressed format (untested):

```cpp
std::unique_ptr<T> temp;
T *src = &v;

if (!v.isCompressed()) {
    temp = make_unique<T>(v);
    temp->makeCompressed();
    src = temp.get();
}

```

Reply:
I chose option 1. Time will tell whether that was a good decision.